forc-test/src/lib.rs

```diff
@@ -365,9 +375,49 @@ impl BuiltTests {
 /// First builds the package or workspace, ready for execution.
 pub fn build(opts: Opts) -> anyhow::Result<BuiltTests> {
     let build_opts = opts.clone().into_build_opts();
-    let built = pkg::build_with_options(build_opts)?;
-    let built_tests = BuiltTests::from_built(built, opts)?;
-    Ok(built_tests)
+
+    let build_plan = pkg::BuildPlan::from_build_opts(&build_opts)?;
+    let manifest_map = build_plan.manifest_map();
+    let mut inject_map = HashMap::new();
+    let mut built_contracts = HashMap::new();
+    for pinned_member in build_plan.member_pinned_pkgs() {
+        let pkg_manifest = manifest_map
+            .get(&pinned_member.id())
+            .ok_or_else(|| anyhow::anyhow!("missing manifest for member to test"))?;
+        let pkg_path = pkg_manifest.dir();
+
+        // If pinned_member is a contract, compile it without tests first to inject CONTRACT_ID
+        // into namespace.
+        if let Ok(TreeType::Contract) = pkg_manifest.program_type() {
+            let build_opts_without_tests = opts
+                .clone()
+                .patch_opts(pkg_path)
+                .into_build_opts()
+                .include_tests(false);
+            let built_contract_without_tests =
+                pkg::build_with_options(build_opts_without_tests)?.expect_pkg()?;
+            let contract_id =
+                pkg::contract_id(&built_contract_without_tests, &fuel_tx::Salt::zeroed());
+
+            built_contracts.insert(pinned_member.name.clone(), built_contract_without_tests);
+
+            // Construct namespace with contract id
+            let contract_id_constant_name = "CONTRACT_ID".to_string();
```

Comment:
Do we have a `const` for the `"CONTRACT_ID"` str literal? I think we also use it in `forc_pkg`, maybe we can import it from there?
```suggestion
            let contract_id_constant_name = "CONTRACT_ID".to_string();
```